### PR TITLE
lib: Initial red black tree implementation

### PR DESCRIPTION
**Pull request by notYuriy**
Not to be merged yet, as `range_alloc.zig` is yet to be patched to use a new library.

### CODE REVIEW

File: src/lib/rbtree.zig [hunk 1]
```diff
@@ -0,0 +1,973 @@
...
+                }
+            }
+        else
+            void,
+        /// kth queries extensions
+        kth: if (enable_kth_queries)
+            struct {
+                /// Fix zero sized pointer issue in zig compiler
+                /// Citation: "the issue is: kth is a zero-size type, so the pointer to kth is itself zero-sized, and therefore has no data"
+                /// We need to make it not zero-sized XD
```
Comment (N00byEdge) [inline src/lib/rbtree.zig:123]: Maybe leave a link to the upstream issue here so that we can see when we can remove it

File: src/lib/rbtree.zig [hunk 2]
```diff
@@ -0,0 +1,973 @@
...
+                    // The solution for the case in which uncle is red is to "push blackness down".
+                    // We recolor parent and uncle to black and grandparent to red.
+                    // It is easy to verify that black heights for all nodes have not changed.
+                    // The only problem we have encountered is that grandparent's parent is red.
+                    // If that is the case, we can have double red again. As such, we continue
+                    // fixing by setting `current` to grandparent
+
+                    // If uncle is red, it is not null
+                    std.debug.assert(uncle != null);
+                    const uncle_nonnull = uncle.?;
```
Comment (N00byEdge) [inline src/lib/rbtree.zig:358]: From this, the case where `uncle` is `null` is UB/unreachable, so the assert above has no effect when you write `uncle.?`

**Review by N00byEdge — CHANGES_REQUESTED** [review]
Everywhere you write `color = Color.Black` etc can actually just be written as `color = .Black`. IMO this is much more readable.

**Comment by N00byEdge** [comment]
In total, great PR, I only had a few issues with this large of a PR, very cool.

File: src/lib/rbtree.zig (hunk 2 above)
Comment (notYuriy) [reply src/lib/rbtree.zig:358]: I initially thought that asserting non null explicitely was a good idea, but then I realized that it is dumb. I deleted most of the asserts like this, but apparently forgot some.

File: src/lib/rbtree.zig (hunk 1 above)
Comment (notYuriy) [reply src/lib/rbtree.zig:123]: Issue in ziglang. I am not sure if there is one, but I will search for it.

**Review by N00byEdge — APPROVED** [review]

**Pull request merged**